### ISSUE TITLE
ZFS kswapd0 activity and RAM usage (probably SPL related)

### ISSUE THREAD

**Issue by mickg10** [zfsonlinux/zfs#3085]
I have an interesting issue:
My server has a 6TB/disk 19-disk RAIDZ3 configuration, and 16GB of ram.
The issue: kswapd0 becomes very active when big files are being copied to/from ZFS.
The root cause - system is out of RAM, I believe.
But, judging from top, sorted by RES (attached below), all userspace is taking up less than 4 GB.

And yet, free tells us (swap is off):

```
             total       used       free     shared    buffers     cached
Mem:      16433308   16149456     283852       7240     132936    1425464
-/+ buffers/cache:   14591056    1842252
Swap:            0          0          0
```

So, the RAM is going somewhere. I was guessing the ARC cache, but judging from my understanding of _/proc/spl/kstat/zfs/arcstats_, it is using 4GB, and cannot use more than 8GB.
The next idea was the dedup table, but looking at zpool status -D , it is at about 360MB in RAM.
I am at this point out of ideas, and any ideas would be welcome :)

Thanks!

mickg

_cat /proc/spl/kstat/zfs/arcstats_

```
5 1 0x01 85 4080 17512406728 263618082559987
name                            type data
hits                            4    161242246
misses                          4    12217883
demand_data_hits                4    2122185
demand_data_misses              4    8214
demand_metadata_hits            4    146310144
demand_metadata_misses          4    5128817
prefetch_data_hits              4    37249
prefetch_data_misses            4    218760
prefetch_metadata_hits          4    12772668
prefetch_metadata_misses        4    6862092
mru_hits                        4    24064855
mru_ghost_hits                  4    1288700
mfu_hits                        4    125886272
mfu_ghost_hits                  4    1715929
deleted                         4    11753671
recycle_miss                    4    423882
mutex_miss                      4    3110
evict_skip                      4    285567370
evict_l2_cached                 4    0
evict_l2_eligible               4    101679654912
evict_l2_ineligible             4    108758787072
hash_elements                   4    483264
hash_elements_max               4    912487
hash_collisions                 4    11432844
hash_chains                     4    145041
hash_chain_max                  4    16
p                               4    1548591616
c                               4    4867860272
c_min                           4    4194304
c_max                           4    8413853696
size                            4    4915757200
hdr_size                        4    153573624
data_size                       4    269224960
meta_size                       4    1205096448
other_size                      4    3287862168
anon_size                       4    149504
anon_evict_data                 4    0
anon_evict_metadata             4    0
mru_size                        4    596536320
mru_evict_data                  4    269224960
mru_evict_metadata              4    18155008
mru_ghost_size                  4    1522345984
mru_ghost_evict_data            4    331615232
mru_ghost_evict_metadata        4    1190730752
mfu_size                        4    877635584
mfu_evict_data                  4    0
mfu_evict_metadata              4    0
mfu_ghost_size                  4    3345394688
mfu_ghost_evict_data            4    2915577856
mfu_ghost_evict_metadata        4    429816832
l2_hits                         4    0
l2_misses                       4    0
l2_feeds                        4    0
l2_rw_clash                     4    0
l2_read_bytes                   4    0
l2_write_bytes                  4    0
l2_writes_sent                  4    0
l2_writes_done                  4    0
l2_writes_error                 4    0
l2_writes_hdr_miss              4    0
l2_evict_lock_retry             4    0
l2_evict_reading                4    0
l2_free_on_write                4    0
l2_abort_lowmem                 4    0
l2_cksum_bad                    4    0
l2_io_error                     4    0
l2_size                         4    0
l2_asize                        4    0
l2_hdr_size                     4    0
l2_compress_successes           4    0
l2_compress_zeros               4    0
l2_compress_failures            4    0
memory_throttle_count           4    0
duplicate_buffers               4    0
duplicate_buffers_size          4    0
duplicate_reads                 4    0
memory_direct_count             4    34870
memory_indirect_count           4    290816
arc_no_grow                     4    0
arc_tempreserve                 4    0
arc_loaned_bytes                4    0
arc_prune                       4    27343
arc_meta_used                   4    4646532240
arc_meta_limit                  4    6310390272
arc_meta_max                    4    7199537424
```

Top output (sorted by RES)

```
  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
 3914 root      39  19 8389060 3.144g   5324 S   0.0 20.1 130:30.24 java
 2250 mysql     20   0  488636  46872   1276 S   0.0  0.3   3:26.04 mysqld
 1974 syslog    20   0  260380  21676    928 S   0.0  0.1   0:38.05 rsyslogd
 1151 root      20   0  332484  20092   2256 S   0.0  0.1   0:00.26 gateone
 3282 root      20   0  282600  12164   6696 S   0.0  0.1   0:10.66 apache2
 2811 root      20   0  413776   8616   1376 S   0.0  0.1   3:35.63 zfswatcher
13841 root      20   0  412724   7940   4576 S   0.0  0.0   3:58.68 smbd
 1905 root      20   0  278708   7248   5292 S   0.0  0.0   0:54.74 smbd
11150 root      20   0  271860   7248   5108 S   0.3  0.0   0:00.03 sshd
30810 root      20   0  273704   7052   4984 S   0.0  0.0   0:00.99 sshd
20741 root      20   0  273852   6964   4896 S   0.0  0.0   0:00.72 sshd
 7333 root      20   0  275012   6752   4664 S   0.0  0.0   0:00.44 sshd
10197 root      20   0   31268   6368   1932 S   0.0  0.0   0:00.27 bash
 3321 www-data  20   0  286820   6268    736 S   0.0  0.0   0:00.01 apache2
 1933 root      20   0  240056   6252   4516 S   0.0  0.0   0:42.12 winbindd
 3324 www-data  20   0  286812   6220    700 S   0.0  0.0   0:00.00 apache2
20530 www-data  20   0  286812   6220    700 S   0.0  0.0   0:00.00 apache2
 3319 www-data  20   0  286820   6076    564 S   0.0  0.0   0:00.00 apache2
 3320 www-data  20   0  286820   6076    564 S   0.0  0.0   0:00.00 apache2
 3323 www-data  20   0  286820   6076    564 S   0.0  0.0   0:00.00 apache2
20529 www-data  20   0  286820   6076    564 S   0.0  0.0   0:00.00 apache2
20531 www-data  20   0  286820   6076    564 S   0.0  0.0   0:00.00 apache2
 7662 root      20   0  280624   6028   3876 S   0.0  0.0   0:10.83 smbd
20836 root      20   0   31140   5912   1600 S   0.0  0.0   0:00.75 bash
10096 root      20   0   24760   5812   1664 S   0.0  0.0   0:00.09 bash
 7387 root      20   0   24744   5800   1620 S   0.0  0.0   0:00.39 bash
30907 root      20   0   24704   5624   1488 S   0.0  0.0   0:00.15 bash
 1902 root      20   0  231692   5508   3920 S   0.0  0.0   0:28.59 winbindd
 1947 root      20   0  249960   5348   3568 S   0.0  0.0   0:00.72 winbindd
11580 root      20   0  282912   5228   3224 S   0.0  0.0   0:00.07 smbd
10095 root      20   0   27128   4444   1240 S   0.0  0.0   0:02.40 tmux
 1935 root      20   0  240160   3964   2348 S   0.0  0.0   0:00.43 winbindd
    1 root      20   0   38620   3636   1380 S   0.0  0.0   0:07.68 init
 1944 root      20   0  231572   3136   1576 S   0.0  0.0   0:00.34 winbindd
 1948 root      20   0  278192   3120   1164 S   0.0  0.0   0:01.07 smbd
 1953 root      20   0  195612   3084   1780 S   0.0  0.0   0:28.54 nmbd
 1490 root      20   0   10232   2596    304 S   0.0  0.0   0:00.01 dhclient
 2145 whoopsie  20   0  337764   2596   1416 S   0.0  0.0   0:00.71 whoopsie
 2106 root      20   0   61364   2584   1908 S   0.0  0.0   1:17.10 sshd
 3033 root      20   0   89944   2448   1636 S   0.0  0.0   3:47.02 vmtoolsd
20738 shellin+  20   0   44144   2424   1696 S   0.0  0.0   0:00.36 ssh
30808 shellin+  20   0   44144   2304   1636 S   0.0  0.0   0:00.11 ssh
```

_zfs list_

```
NAME                         USED  AVAIL  REFER  MOUNTPOINT
tank                        14.9T  66.4T   511K  /tank
tank/deduped                 544G  66.4T   544G  /tank/deduped
tank/disks                  14.4T  66.4T  5.36T  /home
tank/disks/raid_lvm         5.33T  66.4T   435K  /home/raid_lvm
tank/disks/raid_lvm/MICKG   2.57T  66.4T  2.56T  /home/MICKG
tank/disks/raid_lvm/shared  2.75T  66.4T  2.75T  /home/shared
tank/disks/raid_media       3.23T  66.4T  3.23T  /home/media
```

_zpool status -D tank_

```
  pool: tank
 state: ONLINE
status: One or more devices could not be used because the label is missing or
        invalid.  Sufficient replicas exist for the pool to continue
        functioning in a degraded state.
action: Replace the device using 'zpool replace'.
   see: http://zfsonlinux.org/msg/ZFS-8000-4J
  scan: scrub repaired 16K in 19h51m with 0 errors on Fri Feb  6 09:01:10 2015
config:

        NAME                                                 STATE     READ WRITE CKSUM
        tank                                                 ONLINE       0     0     0
          raidz3-0                                           ONLINE       0     0     0
            ata-WDC_WD60EFRX-68MYMN1_WD-WXK1H6420T06         ONLINE       0     0     0
            ata-WDC_WD60EFRX-68MYMN1_WD-WXL1H6441P7Y         ONLINE       0     0     0
            ata-WDC_WD60EFRX-68MYMN1_WD-WXK1H643219A         ONLINE       0     0     0
            ata-WDC_WD60EFRX-68MYMN1_WD-WXL1H645RP4N         ONLINE       3     0     0
            ata-WDC_WD60EFRX-68MYMN1_WD-WXL1H642DVVK         ONLINE       0     0     0
            ata-WDC_WD60EFRX-68MYMN1_WD-WXK1H641W1ED         ONLINE       0     0     0
            ata-WDC_WD60EFRX-68MYMN1_WD-WXP1H640XPHY         ONLINE       0     0     0
            ata-WDC_WD60EFRX-68MYMN1_WD-WXP1H6438Z6S         ONLINE       0     0     0
            ata-WDC_WD60EFRX-68MYMN1_WD-WXP1H642ZY6M         ONLINE       4     0     0
            ata-WDC_WD60EFRX-68MYMN1_WD-WXL1H644JLZE         ONLINE       0     0     0
            ata-WDC_WD60EFRX-68MYMN1_WD-WXP1H641K4WD         ONLINE       0     0     0
            ata-WDC_WD60EFRX-68MYMN1_WD-WXL1H642L0W6         ONLINE       0     0     0
            ata-WDC_WD60EFRX-68MYMN1_WD-WX31D847KC9Z         ONLINE       0     0     0
            ata-WDC_WD60EFRX-68MYMN1_WD-WX21D9421PP0         ONLINE       0     0     0
            ata-WDC_WD60EFRX-68MYMN1_WD-WX21D9421CRT         ONLINE       0     0     0
            ata-WDC_WD60EFRX-68MYMN1_WD-WX21D9421PCX         ONLINE       0     0     0
            ata-WDC_WD60EFRX-68MYMN1_WD-WX11D9435T5D         ONLINE       0     0     0
            ata-WDC_WD60EFRX-68MYMN1_WD-WX11D94354ZU         ONLINE       4     0     0
            ata-WDC_WD60EFRX-68MYMN1_WD-WX11D94355L4         ONLINE       0     0     0
        cache
          scsi-1ATA_INTEL_SSDSA2M160G2GC_CVPO00120029160AGN  UNAVAIL      0     0     0

errors: No known data errors

 dedup: DDT entries 2571320, size 1547 on disk, 156 in core

bucket              allocated                       referenced
______   ______________________________   ______________________________
refcnt   blocks   LSIZE   PSIZE   DSIZE   blocks   LSIZE   PSIZE   DSIZE
------   ------   -----   -----   -----   ------   -----   -----   -----
     1     646K   80.7G   36.1G   41.1G     646K   80.7G   36.1G   41.1G
     2    1.15M    147G   70.0G   78.7G    2.88M    369G    174G    196G
     4     607K   75.9G   40.7G   44.6G    2.84M    363G    195G    213G
     8    63.7K   7.96G   3.69G   4.16G     679K   84.9G   38.9G   44.0G
    16    16.0K   2.00G   1.22G   1.31G     300K   37.5G   22.8G   24.5G
    32      712   88.9M   58.3M   62.6M    27.3K   3.41G   2.21G   2.37G
    64       94   11.8M   6.17M   6.86M    7.70K    985M    532M    588M
   128       39   4.88M   2.43M   2.81M    7.08K    906M    461M    531M
   256       19   2.38M   1.31M   1.44M    7.85K   1005M    556M    610M
   512       24      3M   2.21M   2.30M    17.0K   2.13G   1.54G   1.61G
    1K        1    128K      1K   12.8K    1.04K    133M   1.04M   13.2M
    2K        1    128K      1K   12.8K    3.51K    450M   3.51M   44.9M
 Total    2.45M    314G    152G    170G    7.37M    944G    472G    525G
```

**Comment by mickg10** [comment]
Forgot to mention ZFS and kernel versions:
> dpkg --list | grep zfs
ii  dkms                                                  2.2.0.3-1.1ubuntu5.14.04+zfs9~trusty     
ii  libzfs2                                               0.6.3-5~trusty                          
ii  mountall                                              2.53-zfs1                               
ii  zfs-auto-snapshot                                     1.1.0-0ubuntu1~trusty     
ii  zfs-dkms                                              0.6.3-5~trusty                       
ii  zfs-doc                                               0.6.3-5~trusty                         
ii  zfsutils                                              0.6.3-5~trusty                           
ii  zfswatcher                                            0.03-2                                   


> uname -a
Linux nas64 3.13.0-45-generic #74-Ubuntu SMP Tue Jan 13 19:36:28 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux


> cat /etc/lsb-release
DISTRIB_DESCRIPTION="Ubuntu 14.04.1 LTS"

**Comment by mickg10** [comment]
And another update:
```zfs umount -a``` does not reduce the slab usage.

** *rmmod*ing all zfs modules does reduce usage to 0**

The output of *cat /proc/spl/kmem/slab*

```
--------------------- cache -------------------------------------------------------  ----- slab ------  ---- object -----  --- emergency ---
name                                    flags      size     alloc slabsize  objsize  total alloc   max  total alloc   max  dlock alloc   max
spl_vn_cache                          0x00020         0         0     4096      104      0     0     1      0     0    21      0     0     0
spl_vn_file_cache                     0x00020         0         0     4096      112      0     0     0      0     0     0      0     0     0
spl_zlib_workspace_cache              0x00240         0         0  8388608   268104      0     0     0      0     0     0      0     0     0
ddt_cache                             0x00040    524288    248560   524288    24856      1     1     2     21    10    42      0     0     0
zio_vdev_cache                        0x00040 113246208 100937760  4194304   131088     27    27    51    810   770  1524      0     0     0
zio_buf_20480                         0x00040   1572864    819200   524288    20480      3     3    28     63    40   588     13     0    46
zio_data_buf_20480                    0x00040         0         0   524288    20480      0     0     2      0     0    42      0     0     0
zio_buf_24576                         0x00040   3670016    884736   524288    24576      7     7    35    126    36   630     15     0     8
zio_data_buf_24576                    0x00040         0         0   524288    24576      0     0     2      0     0    36      0     0     0
zio_buf_28672                         0x00040   3145728   1232896  1048576    28672      3     3    17     93    43   527      9     0     6
zio_data_buf_28672                    0x00040         0         0  1048576    28672      0     0     2      0     0    62      0     0     0
zio_buf_32768                         0x00040   3145728   1409024  1048576    32768      3     3    20     84    43   560      9     0    13
zio_data_buf_32768                    0x00040         0         0  1048576    32768      0     0     2      0     0    56      0     0     0
zio_buf_36864                         0x00040   5242880   1437696  1048576    36864      5     5    24    125    39   600     10     0     4
zio_data_buf_36864                    0x00040         0         0  1048576    36864      0     0     2      0     0    50      0     0     0
zio_buf_40960                         0x00040   4194304    655360  1048576    40960      4     4    29     92    16   667     21     0     2
zio_data_buf_40960                    0x00040         0         0  1048576    40960      0     0     2      0     0    46      0     0     0
zio_buf_45056                         0x00040   4194304   1261568  1048576    45056      4     4    38     84    28   798     11     0     1
zio_data_buf_45056                    0x00040         0         0  1048576    45056      0     0     2      0     0    42      0     0     0
zio_buf_49152                         0x00040   4194304   1425408  1048576    49152      4     4    35     76    29   665     13     0     8
zio_data_buf_49152                    0x00040         0         0  1048576    49152      0     0     2      0     0    38      0     0     0
zio_buf_53248                         0x00040   4194304    585728  1048576    53248      4     4    34     72    11   612     18     0     4
zio_data_buf_53248                    0x00040         0         0  1048576    53248      0     0     2      0     0    36      0     0     0
zio_buf_57344                         0x00040   5242880   1777664  1048576    57344      5     5    36     85    31   612     15     0     4
zio_data_buf_57344                    0x00040         0         0  1048576    57344      0     0     2      0     0    34      0     0     0
zio_buf_61440                         0x00040   8388608   2580480  2097152    61440      4     4    22    124    42   682     17     0    10
zio_data_buf_61440                    0x00040         0         0  2097152    61440      0     0     4      0     0   124      0     0     0
zio_buf_65536                         0x00040   4194304   1966080  2097152    65536      2     2    22     60    30   660     14     0     2
zio_data_buf_65536                    0x00040         0         0  2097152    65536      0     0     2      0     0    60      0     0     0
zio_buf_69632                         0x00040   4194304   1601536  2097152    69632      2     2    22     56    23   616     18     0     4
zio_data_buf_69632                    0x00040         0         0  2097152    69632      0     0     2      0     0    56      0     0     0
zio_buf_73728                         0x00040   6291456   1769472  2097152    73728      3     3    25     78    24   650     19     0     3
zio_data_buf_73728                    0x00040         0         0  2097152    73728      0     0     2      0     0    52      0     0     0
zio_buf_77824                         0x00040   4194304   2023424  2097152    77824      2     2    28     50    26   700     19     0    12
zio_data_buf_77824                    0x00040         0         0  2097152    77824      0     0     2      0     0    50      0     0     0
zio_buf_81920                         0x00040   4194304   1884160  2097152    81920      2     2    25     48    23   600     19     0    10
zio_data_buf_81920                    0x00040         0         0  2097152    81920      0     0     2      0     0    48      0     0     0
zio_buf_86016                         0x00040   4194304   1032192  2097152    86016      2     2    28     46    12   644     23     0     2
zio_data_buf_86016                    0x00040         0         0  2097152    86016      0     0     2      0     0    46      0     0     0
zio_buf_90112                         0x00040   4194304    630784  2097152    90112      2     2    35     44     7   770     24     0     4
zio_data_buf_90112                    0x00040         0         0  2097152    90112      0     0     2      0     0    44      0     0     0
zio_buf_94208                         0x00040   6291456   2543616  2097152    94208      3     3    25     63    27   525     25     0     2
zio_data_buf_94208                    0x00040         0         0  2097152    94208      0     0     2      0     0    42      0     0     0
zio_buf_98304                         0x00040   6291456   1179648  2097152    98304      3     3    25     60    12   500     19     0     4
zio_data_buf_98304                    0x00040         0         0  2097152    98304      0     0     2      0     0    40      0     0     0
zio_buf_102400                        0x00040   2097152    409600  2097152   102400      1     1    30     19     4   570     29     0     2
zio_data_buf_102400                   0x00040         0         0  2097152   102400      0     0     2      0     0    38      0     0     0
zio_buf_106496                        0x00040   4194304   1703936  2097152   106496      2     2    29     36    16   522     21     0     5
zio_data_buf_106496                   0x00040         0         0  2097152   106496      0     0     2      0     0    36      0     0     0
zio_buf_110592                        0x00040   2097152    663552  2097152   110592      1     1    30     18     6   540     19     0    13
zio_data_buf_110592                   0x00040         0         0  2097152   110592      0     0     2      0     0    36      0     0     0
zio_buf_114688                        0x00040   4194304   2637824  2097152   114688      2     2    33     34    23   561     16     0    14
zio_data_buf_114688                   0x00040         0         0  2097152   114688      0     0     2      0     0    34      0     0     0
zio_buf_118784                        0x00040         0         0  2097152   118784      0     0     0      0     0     0      0     0     0
zio_data_buf_118784                   0x00040         0         0  2097152   118784      0     0     2      0     0    34      0     0     0
zio_buf_122880                        0x00040         0         0  2097152   122880      0     0     2      0     0    32      0     0     0
zio_data_buf_122880                   0x00040         0         0  2097152   122880      0     0     2      0     0    32      0     0     0
zio_buf_126976                        0x00040         0         0  4194304   126976      0     0     2      0     0    62      0     0     0
zio_data_buf_126976                   0x00040         0         0  4194304   126976      0     0     2      0     0    62      0     0     0
zio_buf_131072                        0x00040  20971520   5898240  4194304   131072      5     5    31    155    45   954      3     0     0
zio_data_buf_131072                   0x00040 3116367872 2803630080  4194304   131072    743   743   762  23156 21390 23607     89   123   321
spill_cache                           0x00040         0         0  4194304   131072      0     0     1      0     0     8      0     0     0
zfs_znode_cache                       0x00020 2318532608 415727144    32768     1064  70756 70756 112974  2051924 390721 3276246      0     0     0
```

**Comment by kernelOfTruth** [comment]
I just also observed kswapd0 climbing to 100% and beyond

make sure you're setting /proc/sys/vm/vfs_cache_pressure

at least to 1000 or above

during the incident it was at 100, setting it higher made kswapd0 disappear (or at least lower its toll on the CPU load)

if I remember correctly - having read that it's generally advisable to have a high vfs_cache_pressure in combination with ZFSonLinux

**Comment by mickg10** [comment]
And another update:
System is stable, if slow, if I limit zfs_arc_max to 3GB (3*1024^3 bytes).
Now there is 2.5GB of cache data as per ```free```, i.e., we are 2.5GB below max ram usage.
And ```arc_adapt``` is working hard (40% of a core) when copying big files.

The *size* line is interesting below. Is this fragmentation?

*cat /proc/spl/kstat/zfs/arcstats | grep -e \^c -e size*
```
c                               4    3136233984
c_min                           4    4194304
c_max                           4    3221225472
size                            4    5019315256
hdr_size                        4    165668200
data_size                       4    0
meta_size                       4    1289566720
other_size                      4    3564080336
anon_size                       4    147456
mru_size                        4    925471232
mru_ghost_size                  4    1471157760
mfu_size                        4    363948032
mfu_ghost_size                  4    1396427776
```

**Comment by soulhunter** [comment]
setting swappiness to 0 helped me.

**Comment by kernelOfTruth** [comment]
@soulhunter interesting !

you ever ran into OOM situations ?

**Comment by soulhunter** [comment]
mmm... no... not really.  I know there was a time when ZFS would eat all your RAM by default, and some people didn't like that, but my on latest deployment, I realized that the maximum ARC cache is now about 50% of the RAM.  I had to increase that, because my use case greatly benefits from disk cache, so, I increased to ~200GB (out of 256G), of course: it ends up using more, but it is ok, it only gets to ~210GB (still observing it, though).
